qiskit/pulse/schedule.py

```diff
@@ -937,54 +736,583 @@ def __eq__(self, other: ScheduleComponent) -> bool:
             This does not check for logical equivalency. Ie.,
 
             ```python
-            >>> (Delay(10)(DriveChannel(0)) + Delay(10)(DriveChannel(0)) ==
-                 Delay(20)(DriveChannel(0)))
+            >>> Delay(10, DriveChannel(0)) + Delay(10, DriveChannel(0))
+                == Delay(20, DriveChannel(0))
             False
             ```
         """
-        channels = set(self.channels)
-        other_channels = set(other.channels)
+        # 0. type check, we consider Instruction is a subtype of schedule
+        if not isinstance(other, (type(self), Instruction)):
+            return False
 
-        # first check channels are the same
-        if channels != other_channels:
+        # 1. channel check
+        if set(self.channels) != set(other.channels):
             return False
 
-        # then verify same number of instructions in each
-        instructions = self.instructions
-        other_instructions = other.instructions
-        if len(instructions) != len(other_instructions):
+        # 2. size check
+        if len(self.instructions) != len(other.instructions):
             return False
 
-        # finally check each instruction in `other` is in this schedule
-        for idx, inst in enumerate(other_instructions):
-            # check assumes `Schedule.instructions` is sorted consistently
-            if instructions[idx] != inst:
+        # 3. instruction check
+        return all(self_inst == other_inst for self_inst, other_inst
+                   in zip(self.instructions, other.instructions))
+
+    def __repr__(self) -> str:
+        name = format(self._name) if self._name else ""
+        instructions = ", ".join([repr(instr) for instr in self.instructions[:50]])
+        if len(self.instructions) > 25:
+            instructions += ", ..."
+        return '{}({}, name="{}")'.format(
+            self.__class__.__name__,
+            instructions,
+            name
+        )
+
+
+def _require_schedule_conversion(function: Callable) -> Callable:
+    """A method decorator to convert schedule block to pulse schedule.
+
+    This conversation is performed for backward compatibility only if all durations are assigned.
+    """
+    @functools.wraps(function)
+    def wrapper(self, *args, **kwargs):
+        from qiskit.pulse.transforms import block_to_schedule
+        if self.is_schedulable():
+            return function(block_to_schedule(self), *args, **kwargs)
+        raise UnassignedDurationError('This method requires all durations to be assigned with '
+                                      'some integer value. Please check `.parameters` to find '
+                                      'unassigned parameter objects.')
+    return wrapper
+
+
+class ScheduleBlock:
```

Comment:
I think implementation is too much details for API docs, because this basically explains the functionality of the class. I needed, I would write `ScheduleBlock` is just a list of instructions with alignment context, which conforms to the pulse builder syntax.